src/bam_to_bed/intron_retention.py

```diff
@@ -0,0 +1,311 @@
+        if cached_data is None:
+            self.overlaps.increment(contig, start, end, name, strand, category)
+        else:
+            cached_contig, cached_start, cached_end, cached_name, cached_strand, cached_category = cached_data
+            if cached_category is Cat.INTRON:
+                self.overlaps.increment(contig, start, end, name, strand, category)
+            elif category is Cat.INTRON:
+                self.overlaps.increment(cached_contig, cached_start, cached_end, cached_name, cached_strand, cached_category)
+
+    def skip_read(self, read):
+        return read.is_secondary or \
+               read.is_duplicate or \
+               read.is_unmapped or \
+               read.is_supplementary or \
+               (read.is_paired and read.mate_is_unmapped)
+
+    def correct_contig(self, contig, handler):
+        try:
+            handler.fetch(contig)
+            return contig
+        except ValueError:
+            return contig.lstrip("chr")
+
+    def calculate(self, contig):
+        with pysam.AlignmentFile(self.bam, mode="rb", threads=self.threads) as bam_handler:
+            with pysam.TabixFile(self.ref, mode="r", parser=pysam.asBed(), threads=self.threads) as ref_handler:
+                contig_ref = self.correct_contig(contig, ref_handler)  # contig can be with or without 'chr' prefix so we need to try to fetch both and see which one works
+                contig_bam = self.correct_contig(contig, bam_handler)
+                intron_iter = ref_handler.fetch(contig_ref)
+                intron = next(intron_iter)                             # get initial value from intron iterator
+                for read in bam_handler.fetch(contig_bam):
+                    if self.skip_read(read):                           # skip all "bad" reads
+                        continue
+                    if read.reference_start - intron.end >= 0:
+                        try:
+                            intron = next(intron_iter)
+                        except StopIteration:
+                            break
+                    else:
+                        intron_data = (
+                            contig,
+                            intron.start,
+                            intron.end,
+                            intron.name,
+                            intron.strand,
+                            self.get_category(read, intron, self.span, self.strandness)
+                        )
+                        if self.paired and read.query_name not in self.cache:
+                            self.cache[read.query_name] = intron_data
+                        else:
+                            self.process_overlaps(
+                                intron_data,
+                                self.cache.get(read.query_name, None)
+                            )
+                            try:
+                                del self.cache[read.query_name]
+                            except KeyError:

```

Comment:
If I'm not mistaken, catching exceptions is always a very slow operation.  Probably worth it to replace with `if read.query_name in self.cache: del self.cache[read.query_name]` unless you think the cache miss will be extremely rare

Reply:
Thanks, I didn't think about it, I was just trying to follow the rule https://docs.quantifiedcode.com/python-anti-patterns/readability/asking_for_permission_instead_of_forgiveness_when_working_with_files.html however, I'll need to check what is more efficient

Reply:
Yeah, this `try...catch` will be slow in case of running the script with single reads, as in this case, we don't put any reads in the `self.cache`

Reply:
Updated. Now we first check for `self.paired` and then skip checking `self.cache'